javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -61,59 +65,73 @@ private static int treeSize(int branchCount, int depthShift) {
     }
 
     static <T> BitMappedTrie<T> ofAll(Object array) {
-        final int size = lengthOf(array);
+        final LeafType<T> type = LeafType.of(array);
+        final int size = type.lengthOf(array);
         if (size == 0) {
             return empty();
         } else {
             int shift = 0;
-            for (; lengthOf(array) > BRANCHING_FACTOR; shift += BRANCHING_BASE) {
-                array = grouped(array, BRANCHING_FACTOR);
+            for (LeafType<T> t = type; t.lengthOf(array) > BRANCHING_FACTOR; shift += BRANCHING_BASE) {
+                array = t.grouped(array, BRANCHING_FACTOR);
+                t = obj();
             }
-            return new BitMappedTrie<>(array, 0, size, shift);
+            return new BitMappedTrie<>(type, array, 0, size, shift);
         }
     }
 
     BitMappedTrie<T> prepend(T leading) {
-        final int newSize = length() + 1;
-        if (length() == 0) {
-            return new BitMappedTrie<>(asArray(leading), offset, newSize, depthShift);
+        if (cannotAdd(leading)) {
```

Comment:
if we can't add an element to the `Vector` (i.e. `null` or `String` to a primitive collection), copy everything once, and add it to that instead.